examples/fluids/qfunctions/shocktube.h

```diff
@@ -0,0 +1,505 @@
+      }
+
+      for (int j=0; j<3; j++)
+        dv[j][0][i] -= wdetJ * nu_shock * drhodx[j];
+
+      for (int k=0; k<3; k++)
+        for (int j=0; j<3; j++)
+          dv[j][k][i] -= wdetJ * nu_shock * du[k][j];
+
+      for (int j=0; j<3; j++)
+        dv[j][4][i] -= wdetJ * nu_shock * dEdx[j];
+    }
+
+    // Stabilization
+
+    // Need the Jacobian for the advective fluxes for stabilization
+    //    indexed as: jacob_F_conv[direction][flux component][solution component]
+    CeedScalar jacob_F_conv[3][5][5] = {{{0.}}};
+    CeedScalar vel_sq = u[0]*u[0] + u[1]*u[1] + u[2]*u[2];
+    CeedScalar rho_sq = rho*rho;
+    for (int j=0; j<3; j++) {
+      jacob_F_conv[j][4][0] = (gamma-1.) * (-u[j]*P/rho - u[j]/rho*(E-vel_sq/
+                                            (2.*rho)) + u[j]*(vel_sq/(2.*rho)));
+      jacob_F_conv[j][4][4] =  u[j] * gamma;
+      for (int k=0; k<3; k++) {
+        jacob_F_conv[j][k+1][0] = -u[j]*u[k] + (j==k?((gamma-1.)*vel_sq/(2*rho_sq)):0.);
+        jacob_F_conv[j][k+1][4] = (j==k?(gamma-1.):0.);
+        jacob_F_conv[j][0][k+1] = (j==k?1.:0.);
+        jacob_F_conv[j][4][k+1] = (j==k?(E/rho + (gamma-1.)/rho*(E-vel_sq/
+                                         (2*rho))):0.) - (gamma-1.)*u[j]*u[k];
+        jacob_F_conv[j][j+1][k+1] = -(gamma-1.)*u[k];
+        jacob_F_conv[j][k+1][k+1] += (j==k?(2*u[j]):(u[j]));
+      }
+    }
+    jacob_F_conv[0][2][1] = u[1];
+    jacob_F_conv[0][3][1] = u[2];
+    jacob_F_conv[1][1][2] = u[0];
+    jacob_F_conv[1][3][2] = u[2];
+    jacob_F_conv[2][1][3] = u[0];
+    jacob_F_conv[2][2][3] = u[1];
+
+    // Transpose of the Jacobian
+    CeedScalar jacob_F_conv_T[3][5][5];
+    for (int j=0; j<3; j++)
+      for (int k=0; k<5; k++)
+        for (int l=0; l<5; l++)
+          jacob_F_conv_T[j][k][l] = jacob_F_conv[j][l][k];
+
+    // dqdx collects drhodx, dUdx and dEdx in one vector
+    CeedScalar dqdx[5][3];
+    for (int j=0; j<3; j++) {
+      dqdx[0][j] = drhodx[j];
+      dqdx[4][j] = dEdx[j];
+      for (int k=0; k<3; k++)
+        dqdx[k+1][j] = dUdx[k][j];
+    }
+
+    // strong_conv = dF/dq * dq/dx    (Strong convection)
+    CeedScalar strong_conv[5] = {0};
+    for (int j=0; j<3; j++)
+      for (int k=0; k<5; k++)
+        for (int l=0; l<5; l++)
+          strong_conv[k] += jacob_F_conv[j][k][l] * dqdx[l][j];
+
+    // Tau elements
+    CeedScalar uX[3];
+    for (int j=0; j<3;
+         j++) uX[j] = dXdx[j][0]*u[0] + dXdx[j][1]*u[1] + dXdx[j][2]*u[2];
+    const CeedScalar uiujgij = uX[0]*uX[0] + uX[1]*uX[1] + uX[2]*uX[2];
+    const CeedScalar Cc      = 1.;
+    const CeedScalar Ce      = 1.;
+    const CeedScalar f1      = rho * sqrt(uiujgij);
+    const CeedScalar TauC    = (Cc * f1) /
+                               (8. * (dXdxdXdxT[0][0] + dXdxdXdxT[1][1] + dXdxdXdxT[2][2]));
+    const CeedScalar TauM    = 1. / (f1>1. ? f1 : 1.);
+    const CeedScalar TauE    = TauM / (Ce * cv);
+
+    const CeedScalar Tau[5]  = {TauC, TauM, TauM, TauM, TauE};
```

Comment:
Could you share a citation for this stabilization formulation? Is it intended for conservative variables? The units are different from what I would expect.